### DISCUSSION TITLE
My code doesn't work, can somebody tell why?

### DISCUSSION THREAD

**Discussion by manen** [recp/cglm#215] [Q&A] [answered]
```c
// main.c
void shaders_set_mat4(uint shader_id, char *name, mat4 val) {
  glUniform4fv(glGetUniformLocation(shader_id, name), 1, val);
}

void main() {
  mat4 trans = {
    1, 0, 0, 0,
    0, 1, 0, 0,
    0, 0, 1, 0,
    0, 0, 0, 1
  };

  shaders_set_mat4(shader_id, "trans", trans);
}
```

```glsl
// vertex.glsl
#version 330 core
layout(location = 0) in vec3 vPos;
layout(location = 1) in vec2 textureCoord;

out vec2 texCoord;

uniform mat4 trans;

void main() {
	gl_Position = vec4((trans * vec4(vPos, 1.0)).xyz, 1.0);
	texCoord = textureCoord;
}

```

I'm pretty sure that's all the code that's relevant. It's all worked without the `trans` variables. Does anyone know what's wrong?

**Comment by shakesoda** [comment, marked as answer]
try `glUniformMatrix4fv` instead

**Reply by manen** [reply]
Oh, thanks :D
but sadly that wasn't it :(

**Comment by recp** [comment]
See https://cglm.readthedocs.io/en/latest/opengl.html

**cglm** matrices and vectors are aligned by default to increase speed if you didn't disable it (see https://cglm.readthedocs.io/en/latest/opt.html), aliged data may cause wrong bytes to transfer OpenGL maybe (size of matrix may not be same as packed data), I'm not sure yet. You can disable alignment or you can copy aligned data to packed temp variable then pass it to OpenGL... 

---

EDIT1:

You can also check the way you build the matrix because the transform matrix itself may be wrong because of wrong order of multiplication...

EDIT2:

You also throw the `w` component after the vector is transformed (by setting w == 1.0 in gl_Position), I guess you are trying to disable perspective division. Maybe you just switch to ortho projection instead? Or it could be:

```glsl
gl_Position = trans * vec4(vPos, 1.0);
```

**Reply by manen** [reply]
hm, thanks.
disabling alignment with `CGLM_ALL_UNALIGNED` lead to the same results sadly.
I'm starting to think, is a diagonal `1.0f` transformation matrix supposed to lead to nothing changing compared to before? All the matrix and vector math is pretty new (meaning i barely don't understand it) to me.
Also, in the docs for OpenGL calls, there are some references to a `count` variable. What should that be? Just asking because currently I'm just passing 1 to `glUniformMatrix4fv`. I also tried just passing a raw array to that function, and it lead to the same nothing appearing on screen. I'm not even sure if this is the good place to ask these tho :D

**Comment by manen** [comment]
Issue solved! If anyone wondering, I failed to realize I had to call `glUseProgram` before calling `glUniformMatrix4fv`.
